Sources/Sentry/SentryProfiler.mm

```diff
@@ -415,26 +462,85 @@ + (void)useFramesTracker:(SentryFramesTracker *)framesTracker
 
 #    pragma mark - Private
 
-+ (void)captureEnvelopeIfFinished:(SentryProfiler *)profiler spanID:(SentrySpanId *)spanID
++ (NSArray *)slicedArray:(NSArray *)array transaction:(SentryTransaction *)transaction
 {
-    [_gProfilersPerSpanID removeObjectForKey:spanID];
-    [profiler->_spansInFlight removeObject:spanID];
-    if (profiler->_spansInFlight.count == 0) {
-        [profiler captureEnvelope];
-        [profiler->_transactions removeAllObjects];
-        SENTRY_LOG_DEBUG(@"Span %@ was last being tracked.", spanID.sentrySpanIdString);
+    const auto firstIndex = [array indexOfObjectPassingTest:^BOOL(
+        NSDictionary<NSString *, id> *_Nonnull sample, NSUInteger idx, BOOL *_Nonnull stop) {
+        const auto absoluteSampleTime =
+            [sample[@"elapsed_since_start_ns"] longLongValue] + _gCurrentProfiler->_startTimestamp;
+        *stop = absoluteSampleTime >= transaction.startSystemTime;
+        return *stop;
+    }];
+
+    if (firstIndex == NSNotFound) {
+        const auto firstSampleAbsoluteTime = [array[0][@"elapsed_since_start_ns"] longLongValue]
```

Comment:
`m`: Can it happen that we have an empty array here? Maybe it makes sense to do an early return in this function if the array is empty. Otherwise, we get index out of bounds here. The same applies for https://github.com/getsentry/sentry-cocoa/blob/56141bd6dfb7d4908a40df537f7f56beb42a83cd/Sources/Sentry/SentryProfiler.mm#L507-L508

Reply:
Good catch. Technically no, in the current implementation–all the callsites currently are gated on checks for > 0 count (see [here](https://github.com/getsentry/sentry-cocoa/blob/56141bd6dfb7d4908a40df537f7f56beb42a83cd/Sources/Sentry/SentryProfiler.mm#L346-L352), [here](https://github.com/getsentry/sentry-cocoa/blob/56141bd6dfb7d4908a40df537f7f56beb42a83cd/Sources/Sentry/SentryProfiler.mm#L412-L413) and [here](https://github.com/getsentry/sentry-cocoa/blob/56141bd6dfb7d4908a40df537f7f56beb42a83cd/Sources/Sentry/SentryProfiler.mm#L380-L383) which are populated after checks [here](https://github.com/getsentry/sentry-cocoa/blob/56141bd6dfb7d4908a40df537f7f56beb42a83cd/Sources/Sentry/SentryMetricProfiler.mm#L98-L110)). But to be safe, I will add an extra check here, since there's no way to enforce it with the compiler in case of future usages.

Reply:
Thanks, I was worried about future changes.